alpa/pipeline_parallel/stage_profiling.py

```diff
@@ -493,115 +570,480 @@ def profile_all(stages, compiled_outputs: Sequence[CompileOutput], meshes,
+    assert len(layers) % 2 == 0
+    num_layers = len(layers) // 2
+    indices = list(range(2 * num_layers))
+    computation_source_ratio = mesh_num_devices / cluster_size
+    is_full_mesh = computation_source_ratio == 1
+    tot_flops = layer_flops_prefix_sum[2 * num_layers]
+    stages = []
+    for start in tqdm.tqdm(range(0, num_layers)):
+        for end in tqdm.tqdm(range(start, num_layers), leave=False):
+            if is_full_mesh and not (start == 0 and end == num_layers - 1):
+                continue
+            flops_ratio = (
+                layer_flops_prefix_sum[end + 1] - layer_flops_prefix_sum[start]
+                + layer_flops_prefix_sum[2 * num_layers - start] -
+                layer_flops_prefix_sum[2 * num_layers - end - 1]) / tot_flops
+            if (computation_source_ratio > flops_ratio *
+                (1 + stage_imbalance_tolerance) or
+                    computation_source_ratio < flops_ratio /
+                (1 + stage_imbalance_tolerance)):
+                continue
+            forward_layer_indices = indices[start:end + 1]
+            backward_layer_indices = indices[2 * num_layers - end -
+                                             1:2 * num_layers - start]
+            selected_apply_grad_layers = [
+                apply_grad_layers[idx]
+                for idx in forward_layer_indices
+                if apply_grad_layers[idx] is not None
+            ]
+            stage_name = f"stage_{start}_{end}"
+            stage_config = generate_stage_info(
+                layers, [forward_layer_indices, backward_layer_indices],
+                accumulator_mapping, acc_grad_outvars, stage_name,
+                selected_apply_grad_layers, apply_grad_global_info)
+            for config_idx, autosharding_config in enumerate(
+                    autosharding_configs):
+                if autosharding_config is not None:
+                    stage_indices = (start, end, mesh_id, config_idx)
+                    stages.append(
+                        (stage_indices, stage_config, autosharding_config))
+    return stages
+
+
+def get_max_n_succ_stages(profile_results: Sequence[ProfileResult]):
+    env = {}
+    initial_size = sum(result.initial_size for result in profile_results)
+    peak_memory = 0
+    available_memory = profile_results[0].available_memory
+    assert all(result.available_memory == available_memory
+               for result in profile_results)
+    n_modules = profile_results[0].n_modules
+    assert n_modules == 2, "Only support forward and backward module"
+    assert all(result.n_modules == n_modules for result in profile_results)
+
+    stage_no = n_modules * len(profile_results)
+    # eliminate_time[var] = k means that the variable can be eliminated after
+    # stage k.
+    eliminate_time = {}
+    required_outvars = set()
+    for i in reversed(range(n_modules)):
+        for stage_result in reversed(profile_results):
```

Comment:
In the backward, you do not need to reverse it because the execution order is reversed. The same applies to the next loop that you need to explicitly reverse the backward part there.

Reply:
That's a good point! Thanks for catching this.